manual/src/cmds/tsan.etex

```diff
@@ -0,0 +1,227 @@
+\chapter{Runtime detection of data races with ThreadSanitizer}
+\label{c:tsan}\cutname{tsan.html}
+%HEVEA\cutname{tsan.html}
+
+\section{s:tsan-overview}{Overview and usage}
+
+OCaml, since version 5.0, allows shared-memory parallelism and thus mutation of
+shared data. This creates the possibility of data races, i.e., unordered
+accesses to the same data (one of them being a write). Data races can be
+dangerous as they are easy to miss and capable of causing unexpected results.
+More information about data races and their consequences can be found in
+section~\ref{s:par_mm_easy} and Chapter~\ref{c:memorymodel}.
+
+To help detect data races, OCaml supports the ThreadSanitizer runtime detector
+since OCaml 5.2. To use ThreadSanitizer (also known as TSan), you must install
+a dedicated version of the compiler configured with \texttt{--enable-tsan}.
+This can be done using opam:
+
+\begin{verbatim}
+opam switch create <YOUR-SWITCH-NAME-HERE> ocaml-option-tsan
+\end{verbatim}
+
+TSan support for OCaml is currently available for the x86_64 architecture, on
+GNU/Linux and macOS systems. Building OCaml with TSan support requires GCC or
+Clang. Minimal supported versions are GCC 11 and Clang 14. Note that TSan data
+race reports with GCC 11 are known to result in poor stack trace reporting (no
+line numbers), which is fixed in GCC 12.
+
+A TSan-enabled compiler differs from a regular compiler in the following way:
+all programs compiled by \texttt{ocamlopt} are instrumented with calls to the
+TSan runtime, and TSan will detect data races encountered during execution.
+
+For instance, consider the following program:
+
+\begin{caml_example*}{verbatim}
+let a = ref 0 and b = ref 0
+
+let d1 () =
+  a := 1;
+  !b
+
+let d2 () =
+  b := 1;
+  !a
+
+let () =
+  let h = Domain.spawn d2 in
+  let r1 = d1 () in
+  let r2 = Domain.join h in
+  assert (not (r1 = 0 && r2 = 0))
+\end{caml_example*}
+
+When compiling and running this program with \texttt{ocamlopt}, you may get
```

Comment:
"When you compiler and run this program with \texttt{ocamlopt}, you may observe data race.."